### PR TITLE
Initial commit for moleculer-mongo-checkers

### PR DESCRIPTION
**Pull request by Vivien-lelouette**

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+version: "3.7"
+
+services:
+  mongo:
+    image: mongo:4.2.2-bionic
+    volumes:
+      - data_mongo:/data/db
```
Comment (gautaz) [inline docker-compose.yml]: As this is just used for testing purposes, I guess there is no need to retain test data in a Docker volume.

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+language: node_js
+sudo: false
+node_js:
+  - 10
+install:
+  - npm i
+before_script:
+  - docker-compose up -d
```
Comment (gautaz) [inline .travis.yml:10]: See https://docs.travis-ci.com/user/docker/, you need at least:

```yaml
services:
  - docker
```

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+version: "3.7"
```
Comment (gautaz) [inline docker-compose.yml:1]: Depending on the docker-compose version available in travis, this is probably a too high version of compose file to ask for, see https://docs.docker.com/compose/compose-file/compose-versioning/ and https://docs.travis-ci.com/user/docker/#using-docker-compose

File: README.md [hunk 4]
```diff
@@ -1,2 +1,25 @@
 # moleculer-mongo-checker
-Mongo checker functions for r2d2bzh/moleculer-healthcheck-middleware
+
+This project provides health checkers for [moleculer-healthcheck-middleware](https://github.com/r2d2bzh/moleculer-healthcheck-middleware) for project using [moleculer-db-adapter-mongo](https://github.com/moleculerjs/moleculer-db/tree/master/packages/moleculer-db-adapter-mongo).
```
Comment (gautaz) [inline README.md]: little typo: "for project**s** using"

File: README.md [hunk 5]
```diff
@@ -1,2 +1,25 @@
...
+
+## Usage
+
+This project gives createChecker functions to use with [moleculer-healthcheck-middleware](https://github.com/r2d2bzh/moleculer-healthcheck-middleware).
+
+**Initialize the Mongo Liveness check**
+
+```js
+const HealthMiddleware = require('@r2d2bzh/moleculer-healthcheck-middleware');
+const HealthcheckMongo = require('moleculer-mongo-healthcheck');
```
Comment (gautaz) [inline README.md]: Do we already want to document the fact that the library will be available on the public npm repo ?

```js
const HealthcheckMongo = require('@r2d2bzh/moleculer-mongo-healthcheck');
```

File: index.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * Creates a Healthcheck checker for Mongo liveness
+ * This must be passed as a parameter to the HealthMiddleware liveness.createChecker
+ * @param {Object} broker to check liveness
+ */
+function createLivenessChecker(broker) {
+  return (next) => {
+    let error;
+    const servicesWithAdapter = broker.services.filter((service) => hasMongoAdapter(service));
```
Comment (gautaz) [inline index.js]: No need to create a lambda, pass directly the `hasMongoAdapter` function:

```js
broker.services.filter(hasMongoAdapter);
```

File: package.json [hunk 7]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "moleculer-healthcheck-mongo-middleware",
+  "version": "0.0.0",
+  "description": "Wrapper of moleculer-healthcheck-middleware to check liveness of Mongo",
+  "main": "index.js",
+  "scripts": {
+    "pretest": "npm install",
+    "test": "ava",
+    "pretest:debug": "npm run pretest",
+    "test:debug": "ava debug --host 0.0.0.0 --port 9229"
```
Comment (gautaz) [inline package.json]: Not sure but I think 9229 is the default port, so no need to overload the default.

File: package.json [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  },
+  "author": "",
+  "license": "ISC",
+  "devDependencies": {
+    "ava": "^3.5.0",
+    "moleculer-repl": "^0.6.3",
+    "nodemon": "^2.0.2",
+    "@r2d2bzh/moleculer-healthcheck-middleware": "https://github.com/vivien-lelouette/moleculer-healthcheck-middleware"
```
Comment (gautaz) [inline package.json]: ok, we must not forget to change this as soon as the healthcheck middleware is available on the public npm.

File: test/index.js [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+      adapter: new MongoDBAdapter('mongodb://localhost'),
+      collection: 'posts',
+    }
+  ]);
+  const port = t.context.healthport;
+  const response = await fetch(`http://127.0.0.1:${port}/live`).catch(error => error);
+  t.deepEqual(response.status, 200);
+});
+
+test('healthcheck answers a 503 when no adapter is found', async (t) => {
```
Comment (gautaz) [inline test/index.js]: I am not sure I understand the goal of this test the right way.
Does this mean that when there is no db adapter to check the services registered to this broker will be considered as dead?

**Review by gautaz — CHANGES_REQUESTED** [review]
A few minor things to fix

File: docker-compose.yml (hunk 3 above)
Comment (Vivien-lelouette) [reply docker-compose.yml:1]: Compose file with version 3.7 supports docker-engine 18.06.0+.
We are now running on the Xenial build dist of Travis.
This build dist has docker-engine in version 18.06.0-ce and docker-compose in version 1.23.1.

The version 3.7 should be the most up-to-date version we can use for now.

File: test/index.js (hunk 9 above)
Comment (Vivien-lelouette) [reply test/index.js]: The goal is to ensure that the liveness check is in error when no adapters are found.
The issue in this case is that in the checker we just give an error message to the next() function.

In this case, the liveness is not working so it passes an error when calling next(). Every error sent by the checker will be sent as 503 by the [moleculer-healthcheck-middleware](https://github.com/r2d2bzh/moleculer-healthcheck-middleware).

If needed, some work needs to be done on [moleculer-healthcheck-middleware](https://github.com/r2d2bzh/moleculer-healthcheck-middleware) on error handling. Passing the message and the error code to next() might be better for this case.

I don't know which is best: keeping the current behavior, changing the next() function of the [moleculer-healthcheck-middleware](https://github.com/r2d2bzh/moleculer-healthcheck-middleware) checker.

File: README.md (hunk 5 above)
Comment (Vivien-lelouette) [reply README.md]: Fixed in https://github.com/Vivien-lelouette/moleculer-mongo-checker/commit/83bff8ae4d01e5734d7d32238d88b092964e7e1c

File: README.md (hunk 4 above)
Comment (Vivien-lelouette) [reply README.md]: Fixed in c4f1d5f

File: .travis.yml (hunk 2 above)
Comment (Vivien-lelouette) [reply .travis.yml:10]: Fixed in f37831822127b31a8a00c0e85993f648037a2e1a

File: docker-compose.yml (hunk 1 above)
Comment (Vivien-lelouette) [reply docker-compose.yml]: Fixed in a1d553443c53a3f75748c741b4a04d0869ec9f61

File: index.js (hunk 6 above)
Comment (Vivien-lelouette) [reply index.js]: Fixed in f0e66af7c5b51cc4003b855362d97b84de9c8e64

File: package.json (hunk 7 above)
Comment (Vivien-lelouette) [reply package.json]: Fixed in 977666d4ab0c7c4fe11f67345642ad83c0411b51

File: package.json (hunk 8 above)
Comment (Vivien-lelouette) [reply package.json]: Done in 7aaff8f0bf1efd712d32b1a83bed58d160778f99

**Review by gautaz — APPROVED** [review]

**Pull request merged**